Whole lotta love.

nicfit opened the pull request:
- PlugREST._put and _delete
- ~~Http methods raise exceptions for error status~~ [PR](https://github.com/bryant/plugdj/pull/2)
- PlugREST - user_info, skip, chat_delete, moderator_set_role +++
- Advance, user leave/join events.
- Python 3 support

The only API breaking change is that PlugEvent only invokes setattr for response["p"].

nicfit committed:
Updates for plug.dj, more rest support, more events, Python3, etc.

bryant commented on plugdj/base.py:
```diff
@@ -1,6 +1,7 @@
-from urlparse import urljoin
+from urllib.parse import urljoin
```
This looks like it would break Python 2 support. That's a pretty big change. Could you instead handle both 2 and 3 here?

bryant commented:
Thanks for working on this. I think it might be too much love contained in a single commit. Would it be possible to split this across different pulls, or at least separate individual commits within this PR? Fine-grained reviews forthcoming.

bryant commented on plugdj/base.py:
```diff
@@ -1,6 +1,7 @@
+from urllib.parse import urljoin
 from requests import Session
 from ws4py.client.threadedclient import WebSocketClient
+from .events import from_json
```
This looks like a bug fix that deserves its pull. Without this, would `SockBase.recv` throw a NameError? If so, that's terrible.

bryant commented on plugdj/base.py:
```diff
@@ -18,12 +19,28 @@ def _post(self, path, return_req=False, **kwargs):
         req = self._session.post(self.to_url(path), **kwargs)
         if return_req:
             return req
+        req.raise_for_status()
```
This would count as a bug fix in its own right. Thanks for catching it. Possible to separate it out?

bryant commented on plugdj/base.py:
```diff
@@ -59,8 +80,8 @@ def room_state(self):
-    def chat_delete(self):
-        raise NotImplemented("chat/")
+    def chat_delete(self, msg_id):
+        return self._delete(f"/chat/{msg_id}")
```
Sorry, but no f-strings, please.

bryant commented on plugdj/base.py:
```diff
@@ -93,8 +114,8 @@ def moderate_mute_user(self, user_id, reason, duration):
-    def moderate_set_role(self):
-        json = {"userID": user_id, "role": role}
+    def moderate_set_role(self, user_id, role):
+        json = {"userID": user_id, "roleID": role}
```
Terrible. Thanks for catching this. Separate bug fix?

bryant commented on plugdj/events.py:
```diff
@@ -1,33 +1,49 @@
+            except KeyError as ex:
                 msg = "malformed event: " + repr(json)
-                raise MalformedEvent, MalformedEvent(msg), trace
+                raise MalformedEvent(msg) from ex
```
This will lose us Py 2.

bryant commented on plugdj/events.py:
```diff
@@ -1,33 +1,49 @@
-            except KeyError:
-                _, _, trace = exc_info()
+                p = json["p"]
+                setattr(self, attr, p[attr] if hasattr(p, "__getitem__") else p)
```
This looks a bit weird. Would you mind explaining a bit here? Thanks in advance.

nicfit replied:
I feel I must have encountered a {"p": <non-map-type>} i.e. not a dict. For now, reverted.

nicfit replied:
Oh, I remember the example now... 

```json
[{"a":"friendRequest","p":"nicfit","s":"dashboard"}]
```

'p' is not a dict, but is is valid json. I guess I'll bring this code back.

bryant commented on plugdj/events.py:
```diff
@@ -37,6 +53,9 @@ def __init__(self, json):
     "ack": AuthAck,
     "chat": Chat,
     "vote": Vote,
+    "advance": Advance,
```
ASCII order please, if you have time.

nicfit committed:
Support python2 again, also added a tox.ini

nicfit committed:
More python2 support.

nicfit committed:
Event order.

nicfit commented:
Hi @bryant ,

This PR is in much better shape now, and passing tests. I added a tox file, the env list can be trimmed, but that works.

I was hoping you did not care about python2, I don't. It works now, but I added six as a dependency.

nicfit commented on plugdj/tests/__init__.py:
```diff
@@ -18,7 +19,7 @@ def test_chat():
 
 def test_send_nonstring_chat():
     assert is_ok(p.join_room(creds.room))
-    p.send_chat(25)
```
I don't really see the need for this to work, but it could be made to do so.

nicfit commented on plugdj/tests/__init__.py:
```diff
@@ -18,7 +19,7 @@ def test_chat():
 def test_send_nonstring_chat():
     assert is_ok(p.join_room(creds.room))
-    p.send_chat(25)
+    tools.assert_raises(TypeError, p.send_chat, 25)
```
I don't really see the need for this to work, but it could be made to do so.

nicfit replied:
@bryant, is there a use-case for sending integers that I'm missing?

bryant replied:
My memory is a little fuzzy here. I think that at the time, I simply wanted to see how the plug server would react to non-string messages. We could XFAIL this for now, if it no longer works.

bryant replied:
My memory is a little fuzzy here. I think that at the time, I simply wanted to see how the plug server would react to non-string messages. We could XFAIL this for now, if it no longer works.

nicfit replied:
Sounds good, leaving it as is seems like the way to go. Anything else you'd like to see frm the PR to get it merged?

bryant replied:
Optimally, we would chunk this into tightly contained patches, or separate PRs altogether. If that seems like a hassle, I could do it (with full credit attributed to you). However, I am a little bit busy these few days. So I will get to this in a couple of days.

nicfit replied:
I'll split things up into sep. PRs

nicfit replied:
First PR of critical bug fixes [here](https://github.com/bryant/plugdj/pull/2). It fixes more than one thing, but it needs merged ASAP in order for the remaining PRs to e submitted, because I want to keep the tests passing as I go/submit.

nicfit committed:
updated ignore
